crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,365 @@
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use crate::quantum_circuit::circuit_data::SliceOrInt::{Int, Slice};
+use crate::quantum_circuit::intern_context::{BitType, IndexType, InternContext};
+use hashbrown::HashMap;
+use pyo3::basic::CompareOp;
+use pyo3::exceptions::{PyIndexError, PyValueError};
+use pyo3::prelude::*;
+use pyo3::types::{IntoPyDict, PySlice};
+use pyo3::{intern, PyObject, PyResult};
+use std::cmp::max;
+use std::iter::zip;
+use std::mem::swap;
+
+#[pyclass(sequence, module = "qiskit._accelerate.quantum_circuit")]
+#[derive(Clone, Debug)]
+pub struct CircuitData {
+    data: Vec<(PyObject, IndexType, IndexType)>,
+    intern_context: Py<InternContext>,
+    new_callable: PyObject,
+    fn_idx_to_qubit: PyObject,
+    fn_idx_to_clbit: PyObject,
+    fn_qubit_to_idx: PyObject,
+    fn_clbit_to_idx: PyObject,
+}
+
+#[derive(FromPyObject)]
+pub struct ElementType {
+    operation: PyObject,
+    qubits: Vec<PyObject>,
+    clbits: Vec<PyObject>,
+}
+
+#[derive(FromPyObject)]
+pub enum SliceOrInt<'a> {
+    Slice(&'a PySlice),
+    Int(isize),
+}
+
+#[pymethods]
+impl CircuitData {
+    #[new]
+    pub fn new(
+        py: Python<'_>,
+        intern_context: Py<InternContext>,
+        new_callable: PyObject,
+        qubits: PyObject,
+        clbits: PyObject,
+        qubit_indices: PyObject,
+        clbit_indices: PyObject,
+    ) -> PyResult<Self> {
+        let fn_get_item = intern!(py, "__getitem__");
+        Ok(CircuitData {
+            new_callable,
+            intern_context,
+            data: Vec::new(),
+            fn_idx_to_qubit: qubits.getattr(py, fn_get_item)?,
+            fn_idx_to_clbit: clbits.getattr(py, fn_get_item)?,
+            fn_qubit_to_idx: qubit_indices.getattr(py, fn_get_item)?,
+            fn_clbit_to_idx: clbit_indices.getattr(py, fn_get_item)?,
+        })
+    }
+
+    pub fn __len__(&self) -> usize {
+        self.data.len()
+    }
+
+    // Note: we also rely on this to make us iterable!
+    pub fn __getitem__(&self, py: Python<'_>, index: SliceOrInt) -> PyResult<PyObject> {
+        match index {
+            Slice(slice) => {
+                let slice = self.convert_py_slice(py, slice)?;
+                let result = slice
+                    .into_iter()
+                    .map(|i| self.__getitem__(py, Int(i)))
+                    .collect::<PyResult<Vec<PyObject>>>()?;
+                Ok(result.into_py(py))
+            }
+            Int(index) => {
+                let index = self.convert_py_index(index, IndexFor::Lookup)?;
+                let extract_args =
+                    |fn_idx_to_bit: &PyObject, args: &Vec<BitType>| -> PyResult<Vec<PyObject>> {
+                        args.iter()
+                            .map(|i| fn_idx_to_bit.call1(py, (*i,)))
+                            .collect()
+                    };
```

Comment:
> We'd also probably be better off here collecting into a PyTuple allocated on the Python heap directly rather than a Rust-heap Vec that will need copying over.

That sounds like a good idea to me, but I can't actually figure out a way to do that without first collecting the elements into a `Vec`. Collecting into a `PyResult<&PyTuple>` or `PyResult<Py<PyTuple>>` doesn't seem to work. The difficulty is that we're collecting a sequence of `Result`s into a single `Result` containing a sequence.

If we were constructing a `PyList`, I could pre-create an empty list and add these items to it in a for-loop. But since `PyTuple` is immutable, I don't think that's an option here.

Reply:
It would certainly be possible with the raw C API, so this may just be a case where if it does turn out to be performance-important, we might want to drop to using the C-API FFI slightly more directly.  I'm fairly sure it can be done in a perfectly safe way, so we might even from to implement the collection trait back up to PyO3 if so.

Reply:
Yeah, it seems like an obvious gap in PyO3 to not have the collection traits for these. Would certainly be a nice contribution upstream 😄.